File: android/src/main/kotlin/net/nend/nendplugin/InterstitialAd.kt
package net.nend.nendplugin

import android.app.Activity
import android.content.Context
import io.flutter.plugin.common.BinaryMessenger
import io.flutter.plugin.common.JSONMethodCodec
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import net.nend.android.NendAdInterstitial
import net.nend.android.NendAdInterstitial.NendAdInterstitialShowResult
import net.nend.android.NendAdInterstitial.NendAdInterstitialStatusCode.FAILED_AD_DOWNLOAD
import net.nend.android.NendAdInterstitial.NendAdInterstitialStatusCode.FAILED_AD_REQUEST
import net.nend.android.NendAdInterstitial.NendAdInterstitialStatusCode.INVALID_RESPONSE_TYPE
import net.nend.android.NendAdInterstitial.NendAdInterstitialStatusCode.SUCCESS
import net.nend.nendplugin.ext.maybeBoolean
import net.nend.nendplugin.ext.maybeInt
import org.json.JSONObject


class InterstitialAd(
    private val activity: Activity?,
    private val context: Context?,
    messenger: BinaryMessenger?
) : MethodChannel.MethodCallHandler, AdBridger(), NendAdInterstitial.OnClickListener {

    override val methodChannel =
        MethodChannel(messenger, "nend_plugin/interstitial", JSONMethodCodec.INSTANCE)

    init {
        methodChannel.setMethodCallHandler(this)
        registerListener()
    }

    override fun onMethodCall(call: MethodCall, result: MethodChannel.Result) {
        when (targetMethod(call.method)) {
            MethodName.LoadAd -> loadAd(call.arguments())
            MethodName.ShowAd -> showAd(call.arguments())
            MethodName.DismissAd -> dismissAd()
            MethodName.EnableAutoReload -> enableAutoReload(call.arguments())
            else -> {
                return result.notImplemented()
            }
        }
        result.success(true)
    }

    private fun registerListener() {
        NendAdInterstitial.setListener { status ->
            when (status) {
                SUCCESS -> {
                    methodChannel.invokeListenerEvent(
                        CallbackName.OnLoaded,
                        mapOf(KEY_RESULT to "LOAD AD SUCCESS")
                    )
                }
                FAILED_AD_DOWNLOAD, INVALID_RESPONSE_TYPE, FAILED_AD_REQUEST -> {
                    methodChannel.invokeListenerEvent(
                        CallbackName.OnFailedToLoad,
                        mapOf(KEY_RESULT to status.name.replace("_", " "))
                    )
                }
                else -> return@setListener
            }
        }
    }

    private fun loadAd(arguments: JSONObject) {
        val adUnit = AdUnit.fromArguments(arguments)
        if (context != null && adUnit != null) {
            adUnit.let { (spotId, apiKey) ->
                NendAdInterstitial.loadAd(context, apiKey, spotId)
            }
        }
    }

    private fun showAd(arguments: JSONObject) {
        val result = activity?.let { activity ->
            arguments.maybeInt("spotId")?.let { spotId ->
                NendAdInterstitial.showAd(activity, spotId, this)
            }
        }

        result?.let { status ->
            when (status) {
                NendAdInterstitialShowResult.AD_SHOW_SUCCESS -> methodChannel.invokeListenerEvent(
                    CallbackName.OnShown,
                    mapOf(KEY_RESULT to "AD SHOW SUCCESS")
                )
                else -> methodChannel.invokeListenerEvent(
                    CallbackName.OnFailedToShow, mapOf(KEY_RESULT to status.name.replace("_", " "))
                )
            }
        }
    }

    private fun dismissAd() {
        NendAdInterstitial.dismissAd()
    }

    private fun enableAutoReload(arguments: JSONObject) {
        arguments.maybeBoolean("enableAutoReload")?.run {
            NendAdInterstitial.isAutoReloadEnabled = this
        }
    }

    override fun onClick(status: NendAdInterstitial.NendAdInterstitialClickType?) {
        when (status) {
            NendAdInterstitial.NendAdInterstitialClickType.DOWNLOAD -> methodChannel.invokeListenerEvent(
                CallbackName.OnAdClicked,
                null
            )
            NendAdInterstitial.NendAdInterstitialClickType.INFORMATION -> methodChannel.invokeListenerEvent(
                CallbackName.OnInformationClicked,
                null
            )
            NendAdInterstitial.NendAdInterstitialClickType.CLOSE -> methodChannel.invokeListenerEvent(
                CallbackName.OnClosed,
                null
            )
            else -> return
        }
    }
}

File: ios/Classes/Interstitial/InterstitialAd.swift
//
//  InterstitialAd.swift
//  nend_plugin
//
//

import UIKit
import Flutter
import NendAd

class InterstitialAd:
    AdBridger,
    NADInterstitialLoadingDelegate,
    NADInterstitialClickDelegate,
    FlutterPlugin
{
    private var registrar: FlutterPluginRegistrar!
    
    override init() {
        super.init()
        NADInterstitial.sharedInstance().clickDelegate = self
        NADInterstitial.sharedInstance().loadingDelegate = self
    }
    
    static func register(with registrar: FlutterPluginRegistrar) {
        let instance = InterstitialAd()
        instance.registrar = registrar
        
        instance.channel = FlutterMethodChannel(
            name: "nend_plugin/interstitial",
            binaryMessenger: registrar.messenger(),
            codec: FlutterJSONMethodCodec.sharedInstance()
        )
        instance.channel.setMethodCallHandler({
            (call: FlutterMethodCall, result: @escaping FlutterResult) -> Void in
            instance.handle(call, result: result)
        })
        registrar.addMethodCallDelegate(instance, channel: instance.channel)
    }
    
    public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        switch targetMethod(call.method) {
            case .showAd:
                showAd(arguments: call.arguments)
            case .loadAd:
                loadAd(arguments: call.arguments)
            case .dismissAd:
                dismissAd()
            case .enableAutoReload:
                enableAutoReload(arguments: call.arguments)
            default:
                return
        }
        result(true)
    }
    
    private func loadAd(arguments: Any?) {
        let adUnit = self.adUnit(argument: arguments)
        let spotId = adUnit.spotId
        let apiKey = adUnit.apiKey
        
        NADInterstitial.sharedInstance().loadAd(withSpotID: spotId, apiKey: apiKey)
    }
    
    private func showAd(arguments: Any?) {
        guard let rootViewController = self.rootViewController else { return }
        let codable = interstitialCodable(argument: arguments)
        guard let spotId = codable.spotId else { return }
        let showResult = NADInterstitial.sharedInstance().showAd(from: rootViewController, spotID: spotId)
        var result: String?
        switch showResult {
            case .AD_SHOW_SUCCESS:
                self.invokeMethod(CallbackName.onShown, arguments: ["result": "AD SHOW SUCCESS"])
            case .AD_SHOW_ALREADY:
                result = "AD SHOW ALREADY"
            case .AD_LOAD_INCOMPLETE:
                result = "AD LOAD INCOMPLETE"
            case .AD_REQUEST_INCOMPLETE:
                result = "AD REQUEST INCOMPLETE"
            case .AD_DOWNLOAD_INCOMPLETE:
                result = "AD DOWNLOAD INCOMPLETE"
            case .AD_FREQUENCY_NOT_REACHABLE:
                result = "AD FREQUENCY NOT REACHABLE"
            case .AD_CANNOT_DISPLAY:
                result = "AD CANNOT DISPLAY"
        }
        
        if result != nil {
            self.invokeMethod(CallbackName.onFailedToShow, arguments: ["result": result])
        }
    }
    
    private func dismissAd() {
        NADInterstitial.sharedInstance().dismissAd()
    }
    
    private func enableAutoReload(arguments: Any?) {
        let codable = interstitialCodable(argument: arguments)
        NADInterstitial.sharedInstance().enableAutoReload = codable.enableAutoReload ?? true
    }
    
    private func interstitialCodable(argument: Any?) -> InterstitialCodable {
        let jsonData = try! JSONSerialization.data(withJSONObject: argument!, options: [])
        return try! JSONDecoder().decode(InterstitialCodable.self, from: jsonData)
    }
    
    func didFinishLoadInterstitialAd(withStatus status: NADInterstitialStatusCode) {
        var result: String?
        switch status {
            case .SUCCESS:
                self.invokeMethod(.onLoaded, arguments: ["result": "LOAD AD SUCCESS"])
            case .INVALID_RESPONSE_TYPE:
                result = "INVALID RESPONSE TYPE"
            case .FAILED_AD_REQUEST:
                result = "FAILED AD REQUEST"
            case .FAILED_AD_DOWNLOAD:
                result = "FAILED AD DOWNLOAD"
            default:
                result = "FAILED AD REQUEST"
        }
        if result != nil {
            self.invokeMethod(.onFailedToLoad, arguments: ["result": result])
        }
    }
    
    func didClick(with type: NADInterstitialClickType) {
        switch type {
            case .DOWNLOAD:
                self.invokeMethod(.onAdClicked, arguments: nil)
            case .INFORMATION:
                self.invokeMethod(.onInformationClicked, arguments: nil)
            case .CLOSE:
                self.invokeMethod(.onClosed, arguments: nil)
        }
    }
}
